File: packages/mobile/modules/board-renderer/android/src/main/java/com/boardsesh/boardrenderer/BoardRendererModule.kt
package com.boardsesh.boardrenderer

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Rect
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer

class BoardRendererModule : Module() {
    private val cacheDir: File by lazy {
        val reactCacheDir = appContext.reactContext?.cacheDir
            ?: throw IllegalStateException("BoardRenderer: reactContext.cacheDir unavailable")
        File(reactCacheDir, "board-thumbnails").also { it.mkdirs() }
    }

    @Volatile
    private var pruned = false

    private fun pruneCacheIfNeeded(maxBytes: Long) {
        val files = cacheDir.listFiles() ?: return
        var totalBytes = files.sumOf { it.length() }
        if (totalBytes <= maxBytes) return

        // Android's File.lastModified() is our LRU proxy — atime is not
        // reliably available on the filesystems Android caches live on.
        val sorted = files.sortedBy { it.lastModified() }
        var removed = 0
        for (file in sorted) {
            if (totalBytes <= maxBytes) break
            val size = file.length()
            if (file.delete()) {
                totalBytes -= size
                removed++
            }
        }
        if (removed > 0) {
            android.util.Log.i("BoardRenderer", "Pruned $removed cached PNGs; new total $totalBytes bytes")
        }
    }

    override fun definition() = ModuleDefinition {
        Name("BoardRenderer")

        AsyncFunction("renderComposite") { configJson: String, backgroundPaths: List<String>, cacheKey: String ->
            if (!pruned) {
                synchronized(this@BoardRendererModule) {
                    if (!pruned) {
                        pruneCacheIfNeeded(CACHE_CAP_BYTES)
                        pruned = true
                    }
                }
            }
            val outputFile = File(cacheDir, "$cacheKey.png")

            if (outputFile.exists()) {
                // Touch mtime so LRU treats hot files as recently used.
                outputFile.setLastModified(System.currentTimeMillis())
                return@AsyncFunction "file://${outputFile.absolutePath}"
            }

            val renderResult = BoardRendererBridge.render(configJson)
                ?: throw Exception("Rust render failed")

            val width = renderResult.width
            val height = renderResult.height
            val rgbaData = renderResult.data

            // Both bitmaps are recycled in finally so a throw between
            // creation and the manual recycle() can't leak native pixel
            // memory (which lives outside the JVM heap and isn't GC'd).
            val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
            var overlayBitmap: Bitmap? = null
            try {
                val canvas = Canvas(bitmap)

                // Draw background images
                for (bgPath in backgroundPaths) {
                    val bgBitmap = BitmapFactory.decodeFile(bgPath)
                    if (bgBitmap != null) {
                        try {
                            canvas.drawBitmap(
                                bgBitmap,
                                null,
                                Rect(0, 0, width, height),
                                null
                            )
                        } finally {
                            bgBitmap.recycle()
                        }
                    }
                }

                // Draw RGBA overlay — tiny-skia returns premultiplied RGBA, and
                // ARGB_8888 bitmaps default to premultiplied storage in the same
                // byte layout, so copyPixelsFromBuffer lets us hand the buffer
                // straight to the bitmap with no per-pixel JVM loop. setPremultiplied
                // is true by default but we set it explicitly so future changes
                // can't accidentally flip it.
                overlayBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
                overlayBitmap.setPremultiplied(true)
                overlayBitmap.copyPixelsFromBuffer(ByteBuffer.wrap(rgbaData))
                canvas.drawBitmap(overlayBitmap, 0f, 0f, null)

                // Encode to PNG
                FileOutputStream(outputFile).use { outputStream ->
                    val written = bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream)
                    if (!written) {
                        outputFile.delete()
                        throw Exception("PNG compression failed")
                    }
                }
            } finally {
                overlayBitmap?.recycle()
                bitmap.recycle()
            }

            "file://${outputFile.absolutePath}"
        }
    }

    private companion object {
        // Cap the on-disk PNG cache so heavy users don't accumulate hundreds
        // of MB of stale renders. The cache lives in context.cacheDir, which
        // Android may also reclaim on its own under storage pressure — this
        // is just our explicit upper bound.
        const val CACHE_CAP_BYTES: Long = 200L * 1024 * 1024
    }
}

File: packages/mobile/modules/board-renderer/ios/BoardRendererModule.swift
import ExpoModulesCore
import UIKit

public class BoardRendererModule: Module {
  // Cap the on-disk PNG cache so heavy users don't accumulate hundreds of
  // MB of stale renders. The cache lives in Library/Caches, which iOS may
  // also reclaim on its own under storage pressure — this is just our
  // explicit upper bound.
  private static let cacheCapBytes: Int = 200 * 1024 * 1024

  private lazy var cacheDir: URL = {
    let dir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
      .appendingPathComponent("board-thumbnails", isDirectory: true)
    do {
      try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    } catch {
      NSLog("[BoardRenderer] Failed to create cache dir at \(dir.path): \(error)")
    }
    return dir
  }()

  // Lazy-initialized once per module lifetime (i.e. once per app launch).
  // Accessing it from renderComposite triggers the prune on first call.
  private lazy var pruneOnce: Void = {
    pruneCacheIfNeeded(maxBytes: BoardRendererModule.cacheCapBytes)
  }()

  private func pruneCacheIfNeeded(maxBytes: Int) {
    // Sort by modificationDate (not contentAccessDate) because that's what
    // we can reliably bump on a cache hit via setAttributes(.modificationDate:).
    // contentAccessDate is read-only via URLResourceKey on most volumes, so
    // mixing the two would let hot files appear stale after a relaunch.
    let keys: [URLResourceKey] = [.contentModificationDateKey, .fileSizeKey]
    guard let entries = try? FileManager.default.contentsOfDirectory(
      at: cacheDir,
      includingPropertiesForKeys: keys,
      options: [.skipsHiddenFiles]
    ) else { return }

    var infos: [(url: URL, size: Int, modified: Date)] = []
    var totalBytes = 0
    for url in entries {
      guard let values = try? url.resourceValues(forKeys: Set(keys)) else { continue }
      let size = values.fileSize ?? 0
      let modified = values.contentModificationDate ?? Date(timeIntervalSince1970: 0)
      infos.append((url, size, modified))
      totalBytes += size
    }

    if totalBytes <= maxBytes { return }

    infos.sort { $0.modified < $1.modified }
    var removed = 0
    for info in infos {
      if totalBytes <= maxBytes { break }
      do {
        try FileManager.default.removeItem(at: info.url)
        totalBytes -= info.size
        removed += 1
      } catch {
        NSLog("[BoardRenderer] Failed to evict \(info.url.lastPathComponent): \(error)")
      }
    }
    if removed > 0 {
      NSLog("[BoardRenderer] Pruned \(removed) cached PNGs; new total \(totalBytes) bytes")
    }
  }

  public func definition() -> ModuleDefinition {
    Name("BoardRenderer")

    AsyncFunction("renderComposite") {
      (configJson: String, backgroundPaths: [String], cacheKey: String) -> String in
      _ = self.pruneOnce
      let outputUrl = self.cacheDir.appendingPathComponent("\(cacheKey).png")

      if FileManager.default.fileExists(atPath: outputUrl.path) {
        // Touch mtime so LRU treats hot files as recently used.
        try? FileManager.default.setAttributes(
          [.modificationDate: Date()], ofItemAtPath: outputUrl.path
        )
        return outputUrl.absoluteString
      }

      let jsonData = Array(configJson.utf8)
      var outData: UnsafeMutablePointer<UInt8>? = nil
      var outLen: UInt32 = 0
      var outWidth: UInt32 = 0
      var outHeight: UInt32 = 0

      let result = jsonData.withUnsafeBufferPointer { buffer in
        guard let baseAddress = buffer.baseAddress else { return Int32(-1) }
        return board_renderer_render(
          baseAddress,
          UInt32(buffer.count),
          &outData,
          &outLen,
          &outWidth,
          &outHeight
        )
      }

      guard result == 0, let pixelData = outData else {
        throw NSError(
          domain: "BoardRenderer", code: Int(result),
          userInfo: [
            NSLocalizedDescriptionKey: "Rust render failed with code \(result)"
          ])
      }

      defer { board_renderer_free(pixelData, outLen) }

      let width = Int(outWidth)
      let height = Int(outHeight)

      // Create composited image using UIGraphicsImageRenderer for
      // automatic memory management and correct color space handling
      let renderer = UIGraphicsImageRenderer(
        size: CGSize(width: width, height: height),
        format: {
          let format = UIGraphicsImageRendererFormat()
          format.scale = 1.0
          format.opaque = false
          return format
        }()
      )

      let pngData = renderer.pngData { rendererContext in
        let context = rendererContext.cgContext

        // CGContext uses bottom-left origin; flip to top-left for UIKit consistency
        context.translateBy(x: 0, y: CGFloat(height))
        context.scaleBy(x: 1, y: -1)

        // Draw background images in order
        for bgPath in backgroundPaths {
          if let bgImage = UIImage(contentsOfFile: bgPath)?.cgImage {
            context.draw(bgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
          }
        }

        // Create CGImage from the RGBA overlay pixel data returned by Rust
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let bitmapInfo = CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue)
        if let overlayContext = CGContext(
          data: UnsafeMutableRawPointer(pixelData),
          width: width,
          height: height,
          bitsPerComponent: 8,
          bytesPerRow: width * 4,
          space: colorSpace,
          bitmapInfo: bitmapInfo.rawValue
        ), let overlayImage = overlayContext.makeImage() {
          context.draw(overlayImage, in: CGRect(x: 0, y: 0, width: width, height: height))
        }
      }

      try pngData.write(to: outputUrl)
      return outputUrl.absoluteString
    }
  }
}
